src/OpenTelemetry.Exporter.Zipkin/Implementation/ZipkinActivityConversionExtensions.cs

```diff
@@ -0,0 +1,219 @@
+
+            string parentId = null;
+            if (activity.ParentSpanId != default)
+            {
+                parentId = EncodeSpanId(activity.ParentSpanId);
+            }
+
+            var attributeEnumerationState = new AttributeEnumerationState
+            {
+                Tags = PooledList<KeyValuePair<string, string>>.Create(),
+            };
+
+            DictionaryEnumerator<string, string, AttributeEnumerationState>.AllocationFreeForEach(activity.Tags, ref attributeEnumerationState, ProcessTagsRef);
+
+            var activitySource = activity.Source;
+            if (!string.IsNullOrEmpty(activitySource.Name))
+            {
+                PooledList<KeyValuePair<string, string>>.Add(ref attributeEnumerationState.Tags, new KeyValuePair<string, string>("library.name", activitySource.Name));
+                if (!string.IsNullOrEmpty(activitySource.Version))
+                {
+                    PooledList<KeyValuePair<string, string>>.Add(ref attributeEnumerationState.Tags, new KeyValuePair<string, string>("library.version", activitySource.Version));
+                }
+            }
+
+            var localEndpoint = defaultLocalEndpoint;
+
+            var serviceName = attributeEnumerationState.ServiceName;
+
+            // override default service name
+            if (!string.IsNullOrWhiteSpace(serviceName))
+            {
+                if (!string.IsNullOrWhiteSpace(attributeEnumerationState.ServiceNamespace))
+                {
+                    serviceName = attributeEnumerationState.ServiceNamespace + "." + serviceName;
+                }
+
+                if (!LocalEndpointCache.TryGetValue(serviceName, out localEndpoint))
+                {
+                    localEndpoint = defaultLocalEndpoint.Clone(serviceName);
+                    LocalEndpointCache.TryAdd(serviceName, localEndpoint);
+                }
+            }
+
+            ZipkinEndpoint remoteEndpoint = null;
+            if ((activity.Kind == ActivityKind.Client || activity.Kind == ActivityKind.Producer) && attributeEnumerationState.RemoteEndpointServiceName != null)
+            {
+                remoteEndpoint = RemoteEndpointCache.GetOrAdd(attributeEnumerationState.RemoteEndpointServiceName, ZipkinEndpoint.Create);
+            }
+
+            var annotations = PooledList<ZipkinAnnotation>.Create();
+            ListEnumerator<ActivityEvent, PooledList<ZipkinAnnotation>>.AllocationFreeForEach(activity.Events, ref annotations, ProcessActivityEventsRef);
+
+            return new ZipkinSpan(
+                EncodeTraceId(context.TraceId, useShortTraceIds),
+                parentId,
+                EncodeSpanId(context.SpanId),
+                ToActivityKind(activity),
+                activity.OperationName,
+                activity.StartTimeUtc.ToEpochMicroseconds(),
+                duration: (long)activity.Duration.TotalMilliseconds * 1000, // duration in Microseconds
```

Comment:
I'm guessing there should be a way to get precise duration instead of a rounded up one?

Reply:
E.g. using Ticks and do math with `TimeSpan.TicksPerMillisecond`.

Reply:
Fixed